akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala

```diff
@@ -540,27 +562,51 @@ private[stream] class ActorGraphInterpreter(_initial: GraphInterpreterShell) ext
     if (activeInterpreters.isEmpty) context.stop(self)
   }
 
+  private def shortCircuitBatch(): Unit = {
+    @tailrec
+    def shortCircuitBatchUtil(): Unit =
+      if (!shortCircuitBuffer.isEmpty) {
+        val input = shortCircuitBuffer.dequeue()
+        processEvent(input)
+        shortCircuitBatchUtil()
```

Comment:
This is an endless loop if it encounters the right circumstances (buffer being refilled while draining it).